File: ServerSample/Nio_Client/src/ServerConnection.java
import java.io.*;
import java.net.InetSocketAddress;
import java.net.Socket;

public abstract class ServerConnection extends Thread {
    //서버의  정보
    private static final String SERVER_IP = "192.168.0.7";
    private static final int SERVER_PORT = 5050;

    //Socket 관련 객체
    private Socket socket;
    protected BufferedReader reader;
    protected BufferedWriter writer;


    public ServerConnection() throws IOException {
        socket = new Socket();
        socket.connect(new InetSocketAddress(SERVER_IP, SERVER_PORT), 2000);
        writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));
        reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
    }

    @Override
    public void run() {
        while (!this.isInterrupted()) {
            receive();
        }
    }

    abstract void send();

    abstract void receive();

    void closeSocket() throws IOException {
        writer.close();
        reader.close();
        socket.close();
    }
}

File: ServerSample/encryptionTest/app/src/main/java/encryption/encryptiontest/Model/ServerConnection.kt
package encryption.encryptiontest.Model

import java.io.*
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketTimeoutException

class ServerConnection : Thread() {
    private val ip: String = "210.205.46.5"
    private val port: Int = 5050

    private val socket: Socket = Socket()
    private lateinit var reader: BufferedReader
    private lateinit var writer: BufferedWriter

    @Throws(
        SocketTimeoutException::class
    )
    override fun run() {
        socket.connect(InetSocketAddress(ip,port),2000)
        writer = BufferedWriter(OutputStreamWriter(socket.getOutputStream()))
        reader = BufferedReader(InputStreamReader(socket.getInputStream()))
    }

    fun sendData(data: String){
        Thread{
            val out = PrintWriter(writer,true)
            out.println(data)
        }.start()
    }
    fun receiveData():String?{
        var result :String?=null
        Thread{
            result = reader.readLine()
        }.start()
        this.join()
        return result
    }
    fun closeClient(){
        reader.close()
        writer.close()
        socket.close()
    }
}